`TestAgentPoolKeepAliveCountUpdatesAgent ` deadlock flakiness

## Failure

#### Link(s) to logs

on master (v19 dev)

- https://github.com/gravitational/teleport/actions/runs/27313179605/job/80687820945?pr=67590

#### Relevant snippet

```
panic: deadlock: all goroutines in bubble are blocked [recovered, repanicked]

goroutine 72 [running]:
testing.tRunner.func1.2({0x50b1660, 0xc000dab908})
	/opt/go/src/testing/testing.go:1974 +0x419
testing.tRunner.func1()
	/opt/go/src/testing/testing.go:1977 +0x67f
panic({0x50b1660?, 0xc000dab908?})
	/opt/go/src/runtime/panic.go:860 +0x13a
internal/synctest.Run(0xc0006bf7e0)
	/opt/go/src/runtime/synctest.go:251 +0x2f4
testing/synctest.Test(0xc000828248, 0x575caa0)
	/opt/go/src/testing/synctest/synctest.go:290 +0x125
github.com/gravitational/teleport/lib/reversetunnel.TestAgentPoolKeepAliveCountUpdatesAgent(0xc000828248)
	/__w/teleport/teleport/lib/reversetunnel/agentpool_test.go:258 +0x2e
testing.tRunner(0xc000828248, 0x575c798)
	/opt/go/src/testing/testing.go:2036 +0x21d
created by testing.(*T).Run in goroutine 1
	/opt/go/src/testing/testing.go:2101 +0xb13

goroutine 119 [sync.WaitGroup.Wait (durable), synctest bubble 1]:
sync.runtime_SemacquireWaitGroup(0x523c101?, 0x80?)
	/opt/go/src/runtime/sema.go:114 +0x2e
sync.(*WaitGroup).Wait(0xc000a0dac0)
	/opt/go/src/sync/waitgroup.go:206 +0xc5
github.com/gravitational/teleport/lib/reversetunnel.(*AgentPool).Stop(0xc000a0d980)
	/__w/teleport/teleport/lib/reversetunnel/agentpool.go:538 +0x53
github.com/gravitational/teleport/lib/reversetunnel.TestAgentPoolKeepAliveCountUpdatesAgent.func1(0xc000bebb08)
	/__w/teleport/teleport/lib/reversetunnel/agentpool_test.go:362 +0x7f7
testing.tRunner(0xc000bebb08, 0x575caa0)
	/opt/go/src/testing/testing.go:2036 +0x21d
created by testing/synctest.testingSynctestTest in goroutine 118
	/opt/go/src/testing/testing.go:2150 +0x5fd

goroutine 118 [chan receive (durable), synctest bubble 1]:
testing/synctest.testingSynctestTest(0xc000828248, 0x575caa0)
	/opt/go/src/testing/testing.go:2151 +0x625
testing/synctest.Test.func1()
	/opt/go/src/testing/synctest/synctest.go:291 +0x45
created by testing/synctest.Test in goroutine 72
	/opt/go/src/testing/synctest/synctest.go:290 +0x125
```

--- a/lib/reversetunnel/agentpool.go
+++ b/lib/reversetunnel/agentpool.go
@@ -265,15 +265,13 @@ func (p *AgentPool) Start() error {
 		"cluster", p.Cluster,
 	)
 
-	p.wg.Add(1)
-	go func() {
+	p.wg.Go(func() {
 		if err := p.run(); err != nil {
 			p.logger.WarnContext(p.ctx, "Agent pool exited", "error", err)
 		}
 
 		p.cancel()
-		p.wg.Done()
-	}()
+	})
 	return nil
 }
 
@@ -296,10 +294,7 @@ func (p *AgentPool) run() error {
 
 			p.logger.Log(p.ctx, level, "Failed to establish reverse tunnel", "error", err)
 		} else {
-			p.wg.Add(1)
-			p.active.add(agent)
-			p.lastConnectivityChange = p.Clock.Now()
-			p.updateConnectedProxies()
+			p.addActiveAgent(p.ctx, agent)
 		}
 
 		err = p.waitForBackoff(p.ctx, p.events)
@@ -309,6 +304,23 @@ func (p *AgentPool) run() error {
 			p.logger.DebugContext(p.ctx, "Failed to wait for backoff", "error", err)
 		}
 	}
+}
+
+// addActiveAgent registers a successfully started agent with the pool.
+func (p *AgentPool) addActiveAgent(ctx context.Context, agent Agent) {
+	p.wg.Add(1)
+	p.active.add(agent)
+	p.lastConnectivityChange = p.Clock.Now()
+
+	if agent.GetState() == AgentClosed {
+		// The agent can close after Start succeeds but before run registers it.
+		// If the AgentClosed callback already ran, it could not remove the agent
+		// from the active set, so reconcile the state after registration.
+		p.handleEvent(ctx, agent)
+		return
+	}
+
+	p.updateConnectedProxies()
 }
 
 // connectAgent connects a new agent and processes any agent events blocking until a

--- a/lib/reversetunnel/agentpool_test.go
+++ b/lib/reversetunnel/agentpool_test.go
@@ -473,14 +473,38 @@ func TestAgentPoolKeepAliveCountUpdatesAgent(t *testing.T) {
 		require.NoError(t, err)
 		pool.tracker.TrackExpected(track.Proxy{Name: "proxy-1"})
 
+		agentClosedEvents := make(chan AgentState, 1)
+		wrapStateCallback := func(agent Agent) AgentStateCallback {
+			callback := pool.getStateCallback(agent)
+			return func(state AgentState) {
+				if state == AgentClosed {
+					select {
+					case agentClosedEvents <- state:
+					default:
+					}
+				}
+				callback(state)
+			}
+		}
+		drainAgentClosedEvents := func() {
+			for {
+				select {
+				case <-agentClosedEvents:
+				default:
+					return
+				}
+			}
+		}
+
 		agentCount := 0
 		var keepAliveCountMax atomic.Int64 // starting at 0 will default to apidefaults.KeepAliveCountMax
 
 		pool.newAgentFunc = func(ctx context.Context, tracker *track.Tracker, lease *track.Lease) (Agent, error) {
 			agentCount++
+			agentNumber := agentCount
 			// Each agent gets a unique principal so it never collides with a
 			// previously claimed proxy.
-			principal := fmt.Sprintf("proxy-%d", agentCount)
+			principal := fmt.Sprintf("proxy-%d", agentNumber)
 
 			mu.Lock()
 			reqs := currentRequests
@@ -524,7 +548,7 @@ func TestAgentPoolKeepAliveCountUpdatesAgent(t *testing.T) {
 			if err != nil {
 				return nil, trace.Wrap(err)
 			}
-			a.stateCallback = pool.getStateCallback(a)
+			a.stateCallback = wrapStateCallback(a)
 			return a, nil
 		}
 
@@ -536,7 +560,6 @@ func TestAgentPoolKeepAliveCountUpdatesAgent(t *testing.T) {
 		}
 
 		require.NoError(t, pool.Start())
-		defer pool.Stop()
 
 		synctest.Wait()
 
@@ -557,10 +580,22 @@ func TestAgentPoolKeepAliveCountUpdatesAgent(t *testing.T) {
 		close(oldReqs)
 
 		synctest.Wait()
+		drainAgentClosedEvents()
 
 		require.Equal(t, 700*time.Millisecond, <-watchdogTimeouts,
 			"second agent: watchdog timeout should be keepAlive(100ms) * keepAliveCount(7)")
 
+		drainAgentClosedEvents()
 		close(newReqs)
+
+		stopDone := make(chan struct{})
+		go func() {
+			pool.Stop()
+			close(stopDone)
+		}()
+
+		synctest.Wait()
+		require.Equal(t, AgentClosed, <-agentClosedEvents)
+		<-stopDone
 	})
 }

